packages/transformer/src/IModelTransformer.ts

```diff
@@ -2939,118 +2948,103 @@ export class IModelTransformer extends IModelExportHandler {
+      : [changedInstanceId];
+    const idsInTargetDb = await Promise.all(
+      idsInSourceDb.map(async (id) => {
+        let identifierValue: string | undefined;
         let element;
-        try {
-          element = this.sourceDb.elements.getElement(id);
-        } catch (err) {
-          return undefined;
+        if (isRelationship) {
+          try {
+            element = this.sourceDb.elements.getElement(id);
+          } catch (err) {
+            return undefined;
+          }
         }
-        const fedGuid = element.federationGuid;
-        let identifierValue: string | undefined;
+        const fedGuid = isRelationship
+          ? element?.federationGuid
+          : change.FederationGuid;
         if (queryCanAccessProvenance) {
-          const aspects: ExternalSourceAspect[] =
-            this.sourceDb.elements.getAspects(
+          for await (const row of this.sourceDb.createQueryReader(
+            `SELECT esa.Identifier FROM bis.ExternalSourceAspect esa WHERE Scope.Id=:scopeId AND Kind=:kind AND Element.Id=:relatedElementId`,
+            QueryBinder.from([
+              this.targetScopeElementId,
+              ExternalSourceAspect.Kind.Element,
               id,
-              ExternalSourceAspect.classFullName
-            ) as ExternalSourceAspect[];
-          for (const aspect of aspects) {
-            if (
-              aspect.element.id === id &&
-              aspect.scope.id === this.targetScopeElementId
-            )
-              identifierValue = aspect.identifier;
-          }
-          if (identifierValue === undefined) {
-            if (mapOfDeletedElemIdToScopeEsas.get(id) !== undefined)
-              identifierValue =
-                mapOfDeletedElemIdToScopeEsas.get(id)!.Identifier;
+            ])
+          )) {
+            identifierValue = row.Identifier;
           }
+          identifierValue =
+            identifierValue ??
+            mapOfDeletedElemIdToScopeEsas.get(id)?.Identifier;
         }
-        return (
+        const targetId =
           (queryCanAccessProvenance && identifierValue) ||
           // maybe batching these queries would perform better but we should
           // try to attach the second db and query both together anyway
-          (fedGuid && this._queryElemIdByFedGuid(this.targetDb, fedGuid))
-        );
-      });
+          (fedGuid && this._queryElemIdByFedGuid(this.targetDb, fedGuid)) ||
+          // FIXME<MIKE>: describe why it's safe to assume nothing has been deleted in provenanceDb
+          // FIXME<NICK>: Is it safe to assume nothing has been deleted in provenanceDb because the provenanceDb is (most likely?) the targetDb if we've made it to this line of code? And we're processing changes
+          // in the context of the sourceDb implying there are no changes in the targetDb to process therefore no deletes in provenanceDb?
+          (!isRelationship && this._queryProvenanceForElement(id));
```

Comment:
so this only occurs when queryCanAccessProvenance is false (assuming `identifierValue` is never undefined in that case, which we should add an assert for), and that is false when the target is the provenanceDb (aka we're in a forward sync). We should make that more explicit imo.

This logic is very complicated so we should probably rewrite it as if/else statements with lots of intermediate variables with long descriptive names. Sorry I didn't do that in the first place.

So yes the provenanceDb is the targetDb, but I need to think about this some more :smile:, we could be forward syncing to a branch that has deleted something. I'm pretty confident that we test this case though, so probably we just need to find that test and follow it in the debugger. I'm just drawing a blank rn.

Reply:
I was wrong (thanks @nick4598), identifierValue is from the ESAs and therefore is regularly undefined in the fedguid case.

Reply:
I moved the queryProvenance out of the local helper function and into each respective element / relationship conditional branch. Now the local helper function is only responsible for checking by fedguid or ESA if the changedata is in the provenancedb